### PR TITLE
Fix tab group sync

### PR DESCRIPTION
**Pull request by 0916dhkim**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [ ] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #0000) and the maintainers have approved on my working plan.

## Motivation

Fixing https://github.com/facebook/docusaurus/issues/8473

## Test Plan

1. Test [Syncing tab choices](https://deploy-preview-8483--docusaurus-2.netlify.app/docs/markdown-features/tabs/#syncing-tab-choices) section of Docusaurus Tabs page. Syncing feature should work as expected.
2. Test [Query string](https://deploy-preview-8483--docusaurus-2.netlify.app/docs/markdown-features/tabs/#query-string) section of the same page. "Android, iOS" tab group should be synced and URL query params should change on select.

### Test links

Deploy preview: https://deploy-preview-8483--docusaurus-2.netlify.app/

## Related issues/PRs

fix https://github.com/facebook/docusaurus/issues/8473

https://github.com/facebook/docusaurus/pull/8225

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx [hunk 1]
```diff
@@ -140,46 +80,21 @@ function TabsComponent(props: Props): JSX.Element {
...
-    tabGroupChoices,
-    setTabGroupChoices,
-  } = useTabGroupChoice();
   const defaultValue =
     defaultValueProp !== undefined
       ? defaultValueProp
       : children.find((child) => child.props.default)?.props.value ??
         children[0]!.props.value;
-
-  const [selectedValue, setSelectedValue] = useState(defaultValue);
```
Comment (0916dhkim) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:154]: Manage `selectedValue` state inside `useTabGroupChoice` hook.

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx [hunk 2]
```diff
@@ -192,11 +107,7 @@ function TabsComponent(props: Props): JSX.Element {
 
     if (newTabValue !== selectedValue) {
       blockElementScrollPositionUntilNextRender(newTab);
-      setSelectedValue(newTabValue);
```
Comment (0916dhkim) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.tsx:195]: I thought of moving `blockElementScrollPositionUntilNextRender` into `tabGroupChoice.tsx`, but decided to leave it as-is. Let me know if you think it's better to move it.

File: packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx [hunk 3]
```diff
@@ -13,61 +13,123 @@ import React, {
...
 
 type ContextValue = {
-  /** A boolean that tells if choices have already been restored from storage */
-  readonly ready: boolean;
-  /** A map from `groupId` to the `value` of the saved choice. */
-  readonly tabGroupChoices: {readonly [groupId: string]: string};
+  /** A map from `groupId` to the `value` of the saved choice in storage. */
+  readonly tabGroupChoicesInStorage: {readonly [groupId: string]: string};
+  /** A map from `searchKey` to the `value` of the choice in query parameter. */
+  readonly tabGroupChoicesInQueryParams: {readonly [searchKey: string]: string};
```
Comment (0916dhkim) [inline packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx:26]: - Remove `ready` property.
- Manage both local storage values & query param values in this context.

File: packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx [hunk 4]
```diff
@@ -13,61 +13,123 @@ import React, {
...
       console.error(err);
     }
-    setReady(true);
   }, []);
 
-  const setTabGroupChoices = useCallback(
-    (groupId: string, newChoice: string) => {
-      setChoices((oldChoices) => ({...oldChoices, [groupId]: newChoice}));
-      setChoiceSyncWithLocalStorage(groupId, newChoice);
+  const setTabGroupChoice = useCallback(
```
Comment (0916dhkim) [inline packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx:101]: `setTabGroupChoice` pushes tab choice to local storage or query param depending on `groupId` & `queryString`.

File: packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx [hunk 5]
```diff
@@ -80,10 +142,56 @@ export function TabGroupChoiceProvider({
...
+    const queryParamValue =
+      searchKey && context.tabGroupChoicesInQueryParams[searchKey];
+    const storageValue = groupId && context.tabGroupChoicesInStorage[groupId];
+    const valueToSync = queryParamValue ?? storageValue;
+    const isValid =
+      !!valueToSync && values.some(({value}) => value === valueToSync);
+    if (isValid && valueToSync !== selectedValue) {
+      setSelectedValue(valueToSync);
+    }
+  }, [
```
Comment (0916dhkim) [inline packages/docusaurus-theme-common/src/contexts/tabGroupChoice.tsx:184]: **Essence of this PR**: Detect changes in local storage or query params, then update `selectedValue` accordingly.

**Review by 0916dhkim — COMMENTED** [review]
# Changes Summary
Current implementation does not update `selectedValue` when there is a query parameter or local storage change (only restores on mount). That's why tab group sync is not working.

Because selected tab is stored in React state, local storage, and query params, I thought it would be helpful to manage all three in `tabGroupChoice.tsx` module.

**Comment by slorber** [comment]
Thanks

Will review this more deeply later, but in the meantime this PR breaks the querystring feature.

Opening this in a new session (clean/empty localstorage) should focus the iOS tab, but it doesn't anymore:
https://deploy-preview-8483--docusaurus-2.netlify.app/docs/markdown-features/tabs/?current-os=ios#query-string

**Comment by slorber** [comment]
Thanks for trying, but I'm going to open a different PR with a totally different way to handle things (useSyncExternalStore)

The current legacy way is not ideal and I'd like to get rid of this provider and react state synchronisation. As we can see, this leads to subtle bugs and is overly complex

**Comment by 0916dhkim** [comment]
@slorber No problem! I was concerned about adding more to the context too, and now excited to see a better implementation :smile:

**Comment by slorber** [comment]
Thanks

Closing this PR in favor of https://github.com/facebook/docusaurus/pull/8486

Can you help review and let me know if you see any issue?

**Pull request closed without merging**